lib/custom-types.js

```diff
@@ -0,0 +1,112 @@
+'use strict';
+var RSVP = require('rsvp');
+var _ = require('lodash');
+
+exports.mapCustomTypes = function(schema, types){
+
+  return _.reduce(Object.keys(schema), mapCustomTypes(schema), []);
+
+  function mapCustomTypes(schema, prefix){
+    return function(memo, key){
+      if((typeof schema[key]).toLowerCase() === "string") {
+        var customType = types[schema[key]];
+        if(!customType) throw new Error("Custom Type \"" + schema[key] + "\" is not defined");
+        memo.push({
+          path: prefix ? [prefix, key].join('.') : key,
+          hooks: customType.hooks,
+          schema: customType.schema,
+          typeId: schema[key],
+          type: customType
+        });
+      } else if (
+        typeof schema[key] === 'object' && //It's an embedded shema
+        _.isArray(schema[key]) && //And IS an array
+        _.isObject(schema[key][0]) && //With schema placed inside
+        ! (
+          schema[key][0].ref || //Which is not a reference configuration
+          schema[key][0].type //Neither it is array of Strings for instance
+        )
+      ) {
```

Comment:
Could we move these conditionals into a named function? It's pretty difficult to follow what's happening!